File: C/Data-Structures/Linked List/linked_list.c
#include<stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};
struct Node* head;
struct Node* get_new_node(int data) {
    // create a new node
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    // assign data to new node
    newNode->data = data;
    // set next as NULL
    newNode->next = NULL;
    // return the new node
    return newNode;
}
void push_back(int data) {
    // Time complexity of push_back is O(N)
    // get a new node with data
    struct Node* newNode = get_new_node(data);
    // if the list is empty then head will be the new node
    if(head == NULL) {
        head = newNode;
        return;
    }
    // if the list is not empty then traverse till the last node
    struct Node* temp = head;
    while(temp->next != NULL) {
        temp= temp->next;
    }
    // point the next field of last node to new node
    // now the new node becomes the last node
    temp->next = newNode;
}
void push_front(int data) {
    // Time complexity of push_front is O(1)
    struct Node* newNode = get_new_node(data);
    newNode->next = head;
    head = newNode;
}
void pop_back() {
    // Time complexity of pop_back is O(N)
    // checking if list is empty
    if(head == NULL) {
        printf("error : can not delete from empty list\n");
        return;
    }
    struct Node* current = head, *prev = NULL;
    // traverse till last element
    while(current->next != NULL) {
        prev = current; // will store the 2nd last node
        current = current->next;
    }
    prev->next = NULL;  // making the 2nd last node as last node
    free(current); // deleting last node
}
void pop_front() {
    // Time complexity of pop_front is O(1)
    // checking if the list is empty
    if(head == NULL) {
        printf("error : can not delete from empty list\n");
        return;
    }
    // temporarily storing the head node
    struct Node* temp = head;
    // changing the head node to next node
    head = head->next;
    // deleting previous head node
    free(temp);
}
int size() {
    // Time complexity of size is O(N)
    int list_size = 0;
    struct Node* temp = head;
    while(temp != NULL) {
        list_size++;
        temp = temp->next;
    }
    return list_size;
}
void insert(int pos, int data) {
    // Time complexity of insert is O(N)
    int list_size = size(); 
    // validating position
    if(pos < 1 || pos > list_size+1) {
        printf("error : invalid position\n");
        return;
    }
    if(pos == 1) {
        push_front(data);
        return;
    }
    if(pos == list_size+1) {
        push_back(data);
        return;
    }
    // traverse till pos - 1 
    struct Node* temp = head;
    for(int i=1; i<pos-1; i++) {
        temp = temp->next;
    }
    struct Node* newNode = get_new_node(data);
    newNode->next = temp->next;
    temp->next = newNode;
}
void erase(int pos) {
    // Time complexity of erase is O(N)
    int list_size = size();
    if(pos < 1 || pos > list_size) {
        printf("error : invalid position\n");
        return;
    }
    if(pos == 1) {
        pop_front();
        return;
    }
    struct Node* temp = head;
    for(int i=1; i<pos-1; i++) {
        temp = temp->next;
    }
    struct Node* nodeTobeDelete = temp->next;
    temp->next = nodeTobeDelete->next;
    free(nodeTobeDelete);
}
void erase_data(int data) {
    // Time complexity of erase_data is O(N)
    struct Node* temp = head;
    if(temp->data == data) {
        pop_front();
        return;
    }
    while(temp->next != NULL && temp->next->data != data) {
        temp = temp->next;
    }
    struct Node* nodeToBeDelete = temp->next;
    temp->next = nodeToBeDelete->next;
    free(nodeToBeDelete);
}
void reverse() {
    // Time complexity of reverse is O(N)
    struct Node* curr = head, *prev = NULL, *next;
    while(curr != NULL) {
        next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    head = prev;
}
void print_list() {
    struct Node* temp = head;
    while(temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}

int main() {
    head = NULL;
    push_back(10);
    push_front(9);
    insert(1, 8);
    insert(4, 11);
    insert(2, 44);
    // pop_back();
    // pop_front();
    printf("size : %d\n", size());
    print_list();
    // erase_data(9);
    // erase(1);
    print_list();
    reverse();
    print_list();

    return 0;
}

File: Java/Algorithms/Insertion Sort/LinkedList.java
public class LinkedList {
	Node head;
	static class Node {
		int data;
		Node next;
		Node(int data) {
			this.data = data;
			next = null;
		}
	}
	LinkedList() {
		head = null;
	}
	public void push_front(int new_data) {
		Node new_node = new Node(new_data);
		new_node.next = head;
		head = new_node;
	}
	public static void insert(LinkedList list, int data) {
		// Create a new node with given data
		Node new_node = new Node(data);

		// If the Linked List is empty,
		// then make the new node as head
		if (list.head == null) {
			list.head = new_node;
		}else {
			// Else traverse till the last node
			// and insert the new_node there
			Node last = list.head;
			while (last.next != null) {
				last = last.next;
			}
			// Insert the new_node at last node
			last.next = new_node;
		}
		// Return the list by head
		// return list;
	}
	public static void printList(LinkedList list) {
		Node currNode = list.head;
		// Traverse through the LinkedList
		while (currNode != null) {
			System.out.print(currNode.data + " ");
			currNode = currNode.next;
		}
		System.out.println();
	}
	public static LinkedList deleteByKey(LinkedList list, int key) {
		Node currNode = list.head, prev = null;
		// CASE 1: If head node itself holds the key to be deleted
		if (currNode != null && currNode.data == key) {
			list.head = currNode.next;
			return list;
		}
		// CASE 2: If the key is somewhere other than at head
		while (currNode != null && currNode.data != key) {
			prev = currNode;
			currNode = currNode.next;
		}
		if (currNode != null) {
			prev.next = currNode.next;
		}
		return list;
	}
	public static LinkedList deleteAtPosition(LinkedList list, int index) {
		Node currNode = list.head, prev = null;
		// CASE 1: head node itself is to be deleted
		if (index == 0 && currNode != null) {
			list.head = currNode.next;
			return list;
		}
		// CASE 2: validating index
		int counter = 0;
		while (currNode != null) {
			if (counter == index) {
				prev.next = currNode.next;
				break;
			}else {
				prev = currNode;
				currNode = currNode.next;
				counter++;
			}
		}
		if (currNode == null) {
			System.out.println("invalid index");
		}
		return list;
	}
	public static LinkedList deleteAtBeginning(LinkedList list) {
		Node currNode = list.head;
		if(currNode == null) {
			System.out.println("error : can not delete from empty list");
			return list;
		}
		list.head = currNode.next;
		return list;
	}
	public static LinkedList deleteAtTail(LinkedList list) {
		Node currNode = list.head;
		if (currNode == null)
            return list;
        if (currNode.next == null) {
			list.head = null;
            return list;
        }
        // Find the second last node
        Node second_last = currNode;
        while (second_last.next.next != null)
            second_last = second_last.next;
        // Change next of second last
        second_last.next = null;
        return list;
	}
	public static void main(String[] args) {
		LinkedList list = new LinkedList();
		insert(list, 10);
		insert(list, 20);
		insert(list, 30);
		insert(list, 40);
		insert(list, 50);
		printList(list);
		list = deleteByKey(list, 10);
		printList(list);
		list = deleteAtBeginning(list);
		printList(list);
		list = deleteAtPosition(list, 0);
		printList(list);
		list = deleteAtTail(list);
		printList(list);
	}
}
